packages/core-data/src/hooks/use-entity-block-editor.js

```diff
@@ -106,7 +113,7 @@ export default function useEntityBlockEditor( kind, name, { id: _id } = {} ) {
 				selection,
 				content: ( { blocks: blocksForSerialization = [] } ) =>
 					__unstableSerializeAndClean( blocksForSerialization ),
-				...updateFootnotesFromMeta( newBlocks, meta ),
+				...onEditEntityRecord( newBlocks, meta ),
```

Comment:
Could this be implemented with a filter? `useEntityBlockEditor` could run all `edits` through a filter before passing them to `editEntityRecord`:
```js
const edits = applyFilters( 'editor.editEntityRecord', { selection, content, blocks, meta } );
```
There is prior art with the `editor.preSavePost` filter in the `editor` package.

Then footnotes could act as a "feature plugin" that adds footnote functionality to various places with hooks.

Reply:
Sure, that works for me. I wanted to propose something with a real code to get the discussion going 😄

Reply:
I'm not sure if using a filter just because of a dependency issue is the right solution. I think "core-data" knowing of "blocks" and "footnotes" is ok, it's actually the right thing to do (core-data knows about the WP entities and their content/structure). What we could get rid of is the dependency towards "block-editor" package which only exists because of the `InnerBlocks.Content` if I'm not wrong.

Reply:
> What we could get rid of is the dependency towards "block-editor" package which only exists because of the InnerBlocks.Content if I'm not wrong.

There are other dependencies involved, which are probably less unwanted:

```js
import {
	getSaveElement,
	__unstableGetBlockProps as getBlockProps,
} from '@wordpress/blocks';
import { RichTextData } from '@wordpress/rich-text';
```

Reply:
The alternative is to reimplement `InnerBlock.Content` and the rest using some primitives mirroring:

https://github.com/WordPress/gutenberg/blob/c20ad9fb2be3cab6e749b1193140acedcb80d81a/packages/block-editor/src/components/inner-blocks/index.js#L306-L312

Reply:
> What we could get rid of is the dependency towards "block-editor" package which only exists because of the `InnerBlocks.Content` if I'm not wrong.

Yes, it's right that `core-data` knows about blocks, footnotes and rich text, the only problem is that it depends on `block-editor` to do that. It should depend only on `blocks` and `rich-text` packages.

And eliminating `block-editor` looks doable. The only function that `core-data` uses from `block-editor` is `getRichTextValues`. That function is implemented using:
1. some helpers from `blocks` and `rich-text`
2. the `RichText.Content` and `InnerBlocks.Content` components
In turn, `RichText.Content` and `InnerBlocks.Content` are very simple components implemented again using helpers from `blocks` and `rich-text`.

The code also relies on the _identity_ of the `RichText.Content` and `InnerBlocks.Content`. These must be globally unique functions, because they are used like:
```jsx
// create JSX element somewhere
const el = <InnerBlocks.Content />;
// inspect the element type somewhere else:
if ( el.type === InnerBlocks.Content ) { /* ... */ }
```

What we could do is:
1. Move `getRichTextValues` from `block-editor` to `blocks`. It's easy because i) it's private API; ii) is used only at one place, in `core-data`; iii) depends only on code from `element`, `blocks` and `rich-text`, which are already dependencies of `blocks`.
2. Move the `RichText.Content` and `InnerBlocks.Content` components also from `block-editor` to `blocks`. Also easy, they also depend only on `blocks` and `rich-text` code.
3. The `RichText` and `InnerBlocks` components in `block-editor` will import and use the `Content` components like this:
   ```js
   import { RichTextContent, InnerBlocksContent } from '@wordpress/blocks';
   RichText.Content = RichTextContent;
   InnerBlocks.Content = InnerBlocksContent;
   ```

That's all. The `block-editor` dependency is eliminated and everything works as before. Also all the three functions/components naturally belong to `blocks` because they are part of the `save` pipeline. And parsing and serializing blocks is one of the main things the `blocks` package does.

The `RichText` and `InnerBlocks` components in `block-editor` have a lot of "edit" logic that belongs to `block-editor` and should stay there. But the `.Content` sub-components which contain the "save" logic can be imported. And anyone who checks the `InnerBlocks.Content` value (from `block-editor`) or the `InnerBlocksContent` value (from `block`) will see the same function. That's important for a lot of the serializing logic (`renderToString` etc.) which uses code like
```js
switch ( el.type ) {
  case InnerBlocks.Content:
    /* ... */
    break;
  case RichText.Content:
    /* ... */
    break;
}
```

How does that sound? @ellatrix should be much more familiar with this code than me, are there any gotchas I'm missing?